plugins/scaffolder/src/extensions/index.tsx

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { JsonValue } from '@backstage/config';
+import { Extension, useApi } from '@backstage/core';
+import { FieldValidation, Field } from '@rjsf/core';
+import { scaffolderApiRef } from '../api';
+
+export type FieldExtensionOptions = {
+  name: string;
+  component: Field;
+  validation: (data: JsonValue, field: FieldValidation) => void;
+};
+
+export function createScaffolderFieldExtension(
+  options: FieldExtensionOptions,
+): Extension<Field> {
+  const WrappingRegister = () => {
+    const scaffolderApi = useApi(scaffolderApiRef);
+    scaffolderApi.registerCustomField(options);
```

Comment:
While this makes things look more simple I think it's abusing the patterns a bit, and I wouldn't want to set this as a precedent. There's a bunch of things that get a bit weird with this, like `registerCustomField` being called multiple times for the same field, there being no restriction of where to put this component extension or to be honest no real reason for its existence at all except that we need some place to call the API methods.

I would rather go with the element method without any field registration through an API, but we can slim it down a bit. There's no need to go full `traverseElementTree` with the different discoverers and whatnot, as we won't be traversing route elements. Should be enough to do a simple `React.Children.map` and collect the component data with that. Maybe we could provide some utility around that in the future that also handles fragments, but prolly just keep it simple for now.

Reply:
Yep ok - this is what I was worried about - makes sense

Reply:
So should go back to storing the state in a context wrapper then rather than the plugin itself?